src/client/InteractionClient.js

```diff
@@ -0,0 +1,235 @@
+  }
+
+  /**
+   * Create a command.
+   * @param {Object} command The command descriptor.
+   * @param {Snowflake} [guildID] Optional guild ID.
+   * @returns {ApplicationCommand} The created command.
+   */
+  async createCommand(command, guildID) {
+    const clientID = this.client.interactionClient.clientID || (await this.client.fetchApplication()).id;
+    let path = this.client.api.applications(clientID);
+    if (guildID) {
+      path = path.guilds(guildID);
+    }
+    const c = await path.commands.post({
+      data: transformCommand(command),
+    });
+    return new ApplicationCommand(this, c, guildID);
+  }
+
+  handle(data) {
+    switch (data.type) {
+      case InteractionType.PING:
+        return {
+          type: InteractionResponseType.PONG,
+        };
+      case InteractionType.APPLICATION_COMMAND: {
+        let timedOut = false;
+        let resolve;
+        const directPromise = new Promise(r => {
+          resolve = r;
+          this.client.setTimeout(() => {
+            timedOut = true;
+            r({
+              type: InteractionResponseType.ACKNOWLEDGE_WITH_SOURCE,
+            });
+          }, 250);
+        });
+
+        const syncHandle = {
+          acknowledge() {
+            if (!timedOut) {
+              resolve({
+                type: InteractionResponseType.DEFERRED_CHANNEL_MESSAGE_WITH_SOURCE,
+              });
+            }
+          },
+          reply(resolved) {
+            if (timedOut) {
+              return false;
+            }
+            resolve({
+              type: InteractionResponseType.CHANNEL_MESSAGE_WITH_SOURCE,
+              data: resolved.data,
+            });
+            return true;
+          },
+        };
+
+        const interaction = new CommandInteraction(this.client, data, syncHandle);
+
+        /**
+         * Emitted when an interaction is created.
+         * @event Client#interactionCreate
+         * @param {Interaction} interaction The interaction which was created.
+         */
+        this.client.emit(Events.INTERACTION_CREATE, interaction);
+
+        return directPromise;
+      }
+      default:
+        throw new RangeError('Invalid interaction data');
+    }
+  }
+
+  /**
+   * An express-like middleware factory which can be used
+   * with webhook interactions.
+   * @returns {Function} The middleware function.
+   */
+  middleware() {
+    return async (req, res) => {
+      const timestamp = req.get('x-signature-timestamp');
+      const signature = req.get('x-signature-ed25519');
+
+      const chunks = [];
+      for await (const chunk of req) {
+        chunks.push(chunk);
+      }
+      const body = Buffer.concat(chunks);
+
+      if (sodium === undefined) {
+        sodium = require('../util/Sodium');
+      }
+      if (
+        !sodium.methods.verify(
+          Buffer.from(signature, 'hex'),
+          Buffer.concat([Buffer.from(timestamp), body]),
+          this.publicKey,
+        )
```

Comment:
are you sure this is correct ? the example shown on the api docs (https://discord.com/developers/docs/interactions/slash-commands#security-and-authorization) is a bit different, and i've tested what you have here and its always failed, the example on the api docs seems to work for me
```suggestion
        !sodium.methods.verify(
          Buffer.from(signature, 'hex'),
          Buffer.from(timestamp + JSON.stringify(body)),
          this.publicKey,
        )
```

Reply:
you should not be passing a parsed body to this function. Just the raw string/buffer you get as the request body.